pkg/controller/controller.go

```diff
@@ -765,6 +764,186 @@ func (c *M3DBController) handleClusterUpdate(cluster *myspec.M3DBCluster) error
+	if err != nil {
+		return fmt.Errorf("error updating statefulset %s: %w", set.Name, err)
+	}
+
+	return nil
+}
+
+func (c *M3DBController) applyStatefulSetUpdate(
+	cluster *myspec.M3DBCluster,
+	actual *appsv1.StatefulSet,
+	expected *appsv1.StatefulSet,
+) (*appsv1.StatefulSet, error) {
+	updated, err := c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).Update(expected)
+	if err != nil {
+		c.logger.Error(err.Error())
+		return nil, err
+	}
+
+	c.logger.Info("updated statefulset",
+		zap.String("name", expected.Name),
+		zap.Int32("actual_readyReplicas", actual.Status.ReadyReplicas),
+		zap.Int32("actual_updatedReplicas", actual.Status.UpdatedReplicas),
+		zap.String("actual_currentRevision", actual.Status.CurrentRevision),
+		zap.String("actual_updateRevision", actual.Status.UpdateRevision),
+		zap.Int32("expected_readyReplicas", expected.Status.ReadyReplicas),
+		zap.Int32("expected_updatedReplicas", expected.Status.UpdatedReplicas),
+		zap.String("expected_currentRevision", expected.Status.CurrentRevision),
+		zap.String("expected_updateRevision", expected.Status.UpdateRevision),
+		zap.Int64("generation", expected.Generation),
+		zap.Int64("observed", expected.Status.ObservedGeneration),
+	)
+	return updated, nil
+}
+
+// updateStatefulSetNodes returns true if it updates any pods
+func (c *M3DBController) updateStatefulSetNodes(
+	cluster *myspec.M3DBCluster,
+	sts *appsv1.StatefulSet,
+) (bool, error) {
+	logger := c.logger.With(
+		zap.String("namespace", cluster.Namespace), zap.String("name", sts.Name),
+	)
+
+	if _, ok := sts.Annotations[annotations.ParallelUpdateInProgress]; !ok {
+		logger.Debug("no update and no rollout in progress so move to next statefulset")
+		return false, nil
+	}
+
+	numPods, err := c.getMaxPodsToUpdate(sts)
+	if err != nil {
+		return false, err
+	}
+
+	if numPods == 0 {
+		return false, errors.New("parallel update annotation set to 0. will not perform pod updates")
+	}
+
+	pods, err := c.podsToUpdate(cluster.Namespace, sts, numPods)
+	if err != nil {
+		return false, err
+	}
+
+	if len(pods) > 0 {
+		names := make([]string, 0, len(pods))
+		for _, pod := range pods {
+			if err := c.kubeClient.CoreV1().
+				Pods(pod.Namespace).
+				Delete(pod.Name, &metav1.DeleteOptions{}); err != nil {
+				return false, err
+			}
+			names = append(names, pod.Name)
+		}
+		logger.Info("restarting pods", zap.Any("pods", names))
+		return true, nil
+	}
+
+	// If there are no pods to update, we're fully rolled out so remove
+	// the update annotation.
+	if err = c.patchStatefulSet(sts, func(set *appsv1.StatefulSet) {
+		delete(set.Annotations, annotations.ParallelUpdateInProgress)
+
+		// NB(nate): K8s handles this for you when using the RollingUpdate update strategy.
+		// However, since OnDelete removes k8s from the pod update process, it's our
+		// responsibility to set this once done rolling out.
+		set.Status.CurrentReplicas = set.Status.UpdatedReplicas
+		set.Status.CurrentRevision = set.Status.UpdateRevision
```

Comment:
This is surprising / also concerning to have to reconcile k8s' status for it. Are there docs somewhere on this?

If we have to do this, I would also prefer we use `Update` here rather than a patch. That way if k8s has modified the statefulset in between when we decided to change this field + when we made the patch, we'll get a conflict and have to re-reconcile.

Reply:
Yeah, it's very unfortunate. I pieced it together from a bug report (https://github.com/kubernetes/kubernetes/issues/73492) and looking at the update code (https://github.com/kubernetes/kubernetes/blob/053e7f4b0430e1f3814bb16bdb506cd59bb26593/pkg/controller/statefulset/stateful_set_utils.go#L377). But definitely let me know if I'm off base here.

Reply:
Also, was looking closer at the StatefulSetInterface and saw this method:

```go
	UpdateStatus(*v1.StatefulSet) (*v1.StatefulSet, error)
```

What are your thoughts on this change then?

```go
	sts.Status.CurrentReplicas = sts.Status.UpdatedReplicas
	sts.Status.CurrentRevision = sts.Status.UpdateRevision

	if sts, err = c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).UpdateStatus(sts); err != nil {
		return false, err
	}

	if err = c.patchStatefulSet(sts, func(set *appsv1.StatefulSet) {
		delete(set.Annotations, annotations.ParallelUpdateInProgress)
	}); err != nil {
		return false, err
	}
```

Would that also generate a conflict since it's no longer patching for the revision update? Also, seems safe to patch to remove the annotation, but correct me if I'm wrong.

Reply:
This SGTM 👍  .